File: app/src/main/java/com/sumian/hw/network/callback/BaseResponseCallback.java
package com.sumian.hw.network.callback;

import android.support.annotation.NonNull;

import com.blankj.utilcode.util.ActivityUtils;
import com.sumian.common.network.error.ErrorCode;
import com.sumian.hw.network.response.ErrorResponse;
import com.sumian.sd.R;
import com.sumian.sd.account.login.LoginActivity;
import com.sumian.sd.app.App;
import com.sumian.sd.utils.JsonUtil;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by jzz
 * on 2017/10/11.
 * desc:
 */

public abstract class BaseResponseCallback<T> implements Callback<T> {

    @Override
    public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
        onFinish();
        if (response.isSuccessful()) {
            T body = response.body();
            onSuccess(body);
        } else {
            ResponseBody errorBody = response.errorBody();
            if (errorBody == null) {
                errorUnknown();
                return;
            }
            try {
                String errorJson = errorBody.string();
                ErrorResponse errorResponse = JsonUtil.fromJson(errorJson, ErrorResponse.class);
                if (errorResponse == null) {
                    errorUnknown();
                } else {
                    int statusCode = response.code();
                    if (statusCode == ErrorCode.UNAUTHORIZED) {
                        onUnauthorized();
                    } else {
                        onFailure(statusCode, errorResponse.getMessage());
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
                errorUnknown();
            }
        }
    }

    protected void onUnauthorized() {
        ActivityUtils.startActivity(LoginActivity.class);
    }

    private void errorUnknown() {
        onFailure(ErrorCode.UNKNOWN, App.Companion.getAppContext().getString(R.string.error_request_failed_hint));
    }

    @Override
    public void onFailure(@NonNull Call<T> call, @NonNull Throwable t) {
        onFinish();
        errorUnknown();
        t.printStackTrace();
    }

    protected abstract void onSuccess(T response);

    protected abstract void onFailure(int code, String message);

    protected void onFinish() {
    }
}

File: app/src/main/java/com/sumian/sd/network/callback/BaseResponseCallback.kt
package com.sumian.sd.network.callback

import com.blankj.utilcode.util.ActivityUtils
import com.blankj.utilcode.util.LogUtils
import com.sumian.hw.network.callback.ErrorCode
import com.sumian.sd.R
import com.sumian.sd.account.login.LoginActivity
import com.sumian.sd.app.App
import com.sumian.sd.network.StatusCode
import com.sumian.sd.network.response.ErrorInfo400
import com.sumian.sd.network.response.ErrorInfo499
import com.sumian.sd.network.response.ErrorResponse
import com.sumian.sd.utils.JsonUtil
import com.sumian.sd.widget.dialog.SumianAlertDialog
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException


/**
 * <pre>
 *     @author : Zhan Xuzhao
 *     e-mail : xuzhao.z@sumian.com
 *     time   : 2018/6/14 10:24
 *     desc   :
 *     version: 1.0
 * </pre>
 */
abstract class BaseResponseCallback<T> : Callback<T> {

    override fun onResponse(call: Call<T>?, response: Response<T>?) {
        onFinish()
        if (response != null && response.isSuccessful) {
            val body = response.body()
            onSuccess(body)
        } else {
            val errorBody = response?.errorBody()
            if (errorBody == null) {
                unknownError()
                return
            }
            try {
                val errorResponse = getErrorResponseFromErrorBody(response.code(), errorBody)
                if (errorResponse == null) {
                    unknownError()
                } else {
                    onFailure(errorResponse.code, errorResponse.message)
                    when (errorResponse.code) {
                        //token 鉴权失败
                        401 -> {
                            ActivityUtils.startActivity(LoginActivity::class.java)
                        }
                        503 -> showSystemIsMaintainDialog()
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
                unknownError()
            }
        }
    }

    private fun unknownError() {
        onFailure(ErrorCode.UNKNOWN, "Error unknown")
    }

    override fun onFailure(call: Call<T>?, t: Throwable?) {
        onFinish()
        LogUtils.d(t?.message)
        t?.let {
            it.printStackTrace()
            if ((it.message === "Socket closed" || it.message === "Canceled")) {
                return
            }
            onFailure(0, App.getAppContext().getString(R.string.error_request_failed_hint))
        }
    }

    private fun getErrorResponseFromErrorBody(code: Int, errorBody: ResponseBody): ErrorResponse? {
        val errorJson = errorBody.string()
        return if (code == StatusCode.BUSINESS_ERROR) {
            val errorInfo = JsonUtil.fromJson(errorJson, ErrorInfo499::class.java)
            ErrorResponse.createFromErrorInfo(errorInfo)
        } else {
            val errorInfo400 = JsonUtil.fromJson(errorJson, ErrorInfo400::class.java)
            ErrorResponse.createFromErrorInfo(errorInfo400)
        }
    }

    private fun showSystemIsMaintainDialog() {
        SumianAlertDialog(ActivityUtils.getTopActivity())
                .setTitle(R.string.system_maintain)
                .setMessage(R.string.system_maintain_desc)
                .setRightBtn(R.string.confirm, null)
                .show()
    }

    protected abstract fun onSuccess(response: T?)

    protected abstract fun onFailure(code: Int, message: String)

    protected open fun onFinish() {}

}